projects/wallet-lib-client-for-event-queue/0003-sequential-tx-processing.md

```diff
@@ -0,0 +1,137 @@
+# Summary
+[summary]: #summary
+
+Process each transaction as they arrive making atomic changes to balance and
+other metadata in storage.
+
+# Motivation
+[motivation]: #motivation
+
+The current history processing strategy is not efficient and is not scalable to
+large numbers of transactions.
+With the new event queue connection we have a solution for the issues that
+required the current aproach so we can now create a more scalable solution.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When the wallet receives a new transaction or an existing transaction is
+updated we need to update the calculated balances for the tokens on the
+transaction since a token may have been spent or a transaction may have been
+voided.
+The current implementation simply erases all wallet metadata (balances, list of
+utxos, etc) and calculates it by iterating over all transactions. While this
+approach was needed in the past, due to how the event system was setup it would
+take longer times with larger wallets.
+
+We will also have to consider the load of processing the history of transactions
+on startup and when we are re-joining the network if the connection goes offline.
+
+## Processing a single transaction
+
+The core of the implementation will be a method to process a single transaction.
+This method will first check if a transaction with the same id is already on
+storage, this means we are updating an existing transaction.
+
+### New transactions
+
+If this is a new transaction but it is voided, we can ignore it and return null.
+This should not happen but we will treat it as a no-op.
+
+Otherwise we need to iterate over the outputs of the transaction and for each one we need
+to check if the address is from our wallet, if it is we save the:
+- tokens involved in the transaction
+- token balance
+- addresses involved in the transaction (only the ones from our wallet)
+- address balance (by token)
+- utxos, for unspent outputs
+- highest address index
+
+We only need to check the outputs since any change of the inputs should be
+calculated by the update of their transaction.
+
+With the information above we can update the necessary data on the storage.
+
+### Update existing transactions
+
+We should first check if the transaction is being voided or unvoided.
+Voiding means the `voided_by` field is `null` on storage and not `null` on the transaction.
+Unvoiding means the `voided_by` field is `null` on the transaction and not `null` on storage.
+
+When voiding a transaction we need to undo any impact it has on the balance of
+the wallet and the other metadata.
+This means calculating the same data as a new transaction but we should make the
+opposite changes (balances will be removed, etc).
+
+When unvoiding a transaction we can treat it as a new transaction since its
+effect on the wallet was already removed when it was voided.
+
+If the `voided_by` field is the same we can iterate on the outputs and calculate
```

Comment:
```suggestion
If the `voided_by` field is empty we can iterate on the outputs and calculate
```

Reply:
Used `null` instead of empty but to the same effect.